### ISSUE TITLE
Could read_all_events_forward miss events?

### ISSUE THREAD

**Issue by woltere** [slashdotdash/eventstore#3]
The read_all_events_forward will miss events on concurrent inserts to the events table.
The values from event_id ar assigned from the sequence on insert. An insert with a higher event_id might be committed and become visible to other transactions before an insert of an event with a lower event_id.

tx 1  INSERT INTO events(...) --  event_id = 1000
tx 2  INSERT INTO events(...) -- event_id = 1001
tx 2  COMMIT
tx 3  SELECT * FROM events WHERE event_id >= 1000 -- returns event with event_id 1001
tx 1 COMMIT
tx 4  SELECT * FROM events WHERE event_id >= 1002 -- returns no events

**Comment by woltere** [comment]
The first solution 

In the second solution I guess all other concurrent inserts would block until the first one finished. Would it not be easier then to set: `SET TRANSACTION ISOLATION LEVEL SERIALIZABLE `? which should have the same effect.
In both cases gap detection would not be needed I think.

For now I'm thinking about this solution:
```
SELECT * FROM events
WHERE event_id > ? AND xmin < txid_snapshot_xmin(txid_current_snapshot()) ORDER BY id;
```
I'm not using Elixir, I'm looking how different implementations of eventstores on postgresql solve this.

**Comment by slashdotdash** [comment]
@woltere  [Marten](https://github.com/JasperFx/marten) is another PostgreSQL-based event store currently being developed (in .NET) that you might be interested in looking at. They're using stored procedures for event appending: https://github.com/JasperFx/marten/blob/master/src/Marten/Schema/SQL/mt_stream.sql

**Comment by slashdotdash** [comment]
Work in progress is on the [single-writer](https://github.com/slashdotdash/eventstore/tree/single-writer) branch. I'm reworking the event storage to go through a single writer process. That will assign a unique, monotonically incrementing event id then do the insert into the events table.

This is a significant internal change, but there should be no impact to the consumer API.

**Comment by stuartc** [comment]
Interesting change, make sense. Curious to consider the implications of a multi-node application would be; I'm not using this at the moment - I need to look at the writer implementation still; 
have you considered gproc for registering process information?
Would that have any value in tracking writes/stream increments?

**Comment by slashdotdash** [comment]
There's an existing process registry for logical event streams ([streams.ex](https://github.com/slashdotdash/eventstore/blob/master/lib/event_store/streams.ex)). This is used to lookup the pid, or start a new process, when [appending to a stream](https://github.com/slashdotdash/eventstore/blob/master/lib/event_store.ex#L39). 

Have looked at gproc, but not yet used. Not yet convinced of benefits of using it over the registry linked above. But open to suggestions to improve.

**Comment by slashdotdash** [comment]
Fixed by #9.
